File: vda/common/name_fmt.py
VDA_PREFIX = "vda"

BACKEND_NQN_PREFIX = "nqn.2016-06.io.spdk:vda-be"
FRONTEND_NQN_PREFIX = "nqn.2016-06.io.spdk:vda-fe"
EXP_NQN_PREFIX = "nqn.2016-06.io.spdk:vda-exp"
SECONDARY_NQN_PREFIX = "nqn.2016-06.io.spdk:vda-se"

PD_BDEV_TYPE = "000"
PD_LVS_TYPE = "001"
BACKEND_BDEV_TYPE = "002"
FRONTEND_NVME_TYPE = "003"
AGG_BDEV_TYPE = "004"
DA_LVS_TYPE = "005"
SECONDARY_NVME_TYPE = "006"
GRP_BDEV_TYPE = "007"

RAID0_BDEV_TYPE = "100"


class NameFmt:

    @classmethod
    def pd_bdev_prefix(cls):
        return f"{VDA_PREFIX}-{PD_BDEV_TYPE}"

    @classmethod
    def pd_lvs_prefix(cls):
        return f"{VDA_PREFIX}-{PD_LVS_TYPE}"

    @classmethod
    def backend_lvol_prefix(cls):
        return f"{VDA_PREFIX}-{PD_LVS_TYPE}"

    @classmethod
    def backend_nqn_prefix(cls):
        return f"{BACKEND_NQN_PREFIX}"

    @classmethod
    def frontend_nqn_prefix(cls):
        return f"{FRONTEND_NQN_PREFIX}"

    @classmethod
    def frontend_nvme_prefix(cls):
        return f"{VDA_PREFIX}-{FRONTEND_NVME_TYPE}"

    @classmethod
    def agg_bdev_prefix(cls):
        return f"{VDA_PREFIX}-{AGG_BDEV_TYPE}"

    @classmethod
    def da_lvs_prefix(cls):
        return f"{VDA_PREFIX}-{DA_LVS_TYPE}"

    @classmethod
    def secondary_nvme_prefix(cls):
        return f"{VDA_PREFIX}-{SECONDARY_NVME_TYPE}"

    @classmethod
    def grp_bdev_prefix(cls):
        return f"{VDA_PREFIX}-{GRP_BDEV_TYPE}"

    @classmethod
    def raid0_bdev_prefix(cls):
        return f"{VDA_PREFIX}-{RAID0_BDEV_TYPE}"

    @classmethod
    def exp_nqn_prefix(cls):
        return f"{EXP_NQN_PREFIX}"

    @classmethod
    def secondary_nqn_prefix(cls):
        return f"{SECONDARY_NQN_PREFIX}"

    @classmethod
    def pd_bdev_name(cls, pd_id):
        prefix = cls.pd_bdev_prefix()
        return f"{prefix}-{pd_id}"

    @classmethod
    def pd_lvs_name(cls, pd_id):
        prefix = cls.pd_lvs_prefix()
        return f"{prefix}-{pd_id}"

    @classmethod
    def backend_lvol_name(cls, vd_id):
        return f"{vd_id}"

    @classmethod
    def backend_lvol_fullname(cls, pd_id, vd_id):
        lvs_name = cls.pd_lvs_name(pd_id)
        lvol_name = cls.backend_lvol_name(vd_id)
        return f"{lvs_name}/{lvol_name}"

    @classmethod
    def backend_nqn_name(cls, vd_id):
        prefix = cls.backend_nqn_prefix()
        return f"{prefix}-{vd_id}"

    @classmethod
    def frontend_nqn_name(cls, cntlr_id):
        prefix = cls.frontend_nqn_prefix()
        return f"{prefix}-{cntlr_id}"

    @classmethod
    def frontend_nvme_name(cls, vd_id):
        prefix = cls.frontend_nvme_prefix()
        return f"{prefix}-{vd_id}"

    @classmethod
    def agg_bdev_name(cls, da_id):
        prefix = cls.agg_bdev_prefix()
        return f"{prefix}-{da_id}"

    @classmethod
    def da_lvs_name(cls, da_id):
        prefix = cls.da_lvs_prefix()
        return f"{prefix}-{da_id}"

    @classmethod
    def raid0_bdev_name(cls, grp_id):
        prefix = cls.raid0_bdev_prefix()
        return f"{prefix}-{grp_id}"

    @classmethod
    def grp_bdev_name(cls, grp_id):
        prefix = cls.grp_bdev_prefix()
        return f"{prefix}-{grp_id}"

    @classmethod
    def main_snap_name(cls, da_id):
        return f"{da_id}"

    @classmethod
    def snap_full_name(cls, da_id, snap_name):
        da_lvs_name = cls.da_lvs_name(da_id)
        if not snap_name:
            snap_name = cls.main_snap_name(da_id)
        return f"{da_lvs_name}/{snap_name}"

    @classmethod
    def exp_nqn_name(cls, da_name, exp_name):
        prefix = cls.exp_nqn_prefix()
        return f"{prefix}-{da_name}-{exp_name}"

    @classmethod
    def secondary_nqn_name(cls, cntlr_id):
        prefix = cls.secondary_nqn_prefix()
        return f"{prefix}-{cntlr_id}"

    @classmethod
    def secondary_nvme_name(cls, primary_cntlr_id):
        prefix = cls.secondary_nvme_prefix()
        return f"{prefix}-{primary_cntlr_id}"

File: pkg/lib/namefmt.go
package lib

import (
	"fmt"
	"strings"
)

const (
	pdBdevType    = "000"
	pdLvsType     = "001"
	vdBeBdevType  = "002"
	vdFeNvmeType  = "003"
	aggBdevType   = "004"
	daLvsType     = "005"
	secNvmeType   = "006"
	grpBdevType   = "007"
	vdSusresType  = "008"
	mtNullType    = "009"
	mtConcatType  = "010"
	mtRaid1Type   = "011"
	raid0BdevType = "100"
	raid1BdevType = "101"

	beNqnPrefix  = "be"
	feNqnPrefix  = "fe"
	expNqnPrefix = "exp"
	secNqnPrefix = "sec"
)

type NameFmt struct {
	vdaPrefix string
	nqnPrefix string
}

func (nf *NameFmt) PdBdevPrefix() string {
	return fmt.Sprintf("%s-%s", nf.vdaPrefix, pdBdevType)
}

func (nf *NameFmt) PdLvsPrefix() string {
	return fmt.Sprintf("%s-%s", nf.vdaPrefix, pdLvsType)
}

func (nf *NameFmt) BeLvolPrefix() string {
	return fmt.Sprintf("%s-%s", nf.vdaPrefix, pdLvsType)
}

func (nf *NameFmt) BeNqnPrefix() string {
	return fmt.Sprintf("%s:%s", nf.nqnPrefix, beNqnPrefix)
}

func (nf *NameFmt) FeNqnPrefix() string {
	return fmt.Sprintf("%s:%s", nf.nqnPrefix, feNqnPrefix)
}

func (nf *NameFmt) FeNvmePrefix() string {
	return fmt.Sprintf("%s:%s", nf.vdaPrefix, vdFeNvmeType)
}

func (nf *NameFmt) AggBdevPrefix() string {
	return fmt.Sprintf("%s-%s", nf.vdaPrefix, aggBdevType)
}

func (nf *NameFmt) DaLvsPrefix() string {
	return fmt.Sprintf("%s-%s", nf.vdaPrefix, daLvsType)
}

func (nf *NameFmt) SnapFullNamePrefix() string {
	return nf.DaLvsPrefix()
}

func (nf *NameFmt) SecNvmePrefix() string {
	return fmt.Sprintf("%s-%s", nf.vdaPrefix, secNvmeType)
}

func (nf *NameFmt) GrpBdevPrefix() string {
	return fmt.Sprintf("%s-%s", nf.vdaPrefix, grpBdevType)
}

func (nf *NameFmt) Raid0BdevPrefix() string {
	return fmt.Sprintf("%s-%s", nf.vdaPrefix, raid0BdevType)
}

func (nf *NameFmt) Raid1BdevPrefix() string {
	return fmt.Sprintf("%s-%s", nf.vdaPrefix, raid1BdevType)
}

func (nf *NameFmt) ExpNqnPrefix() string {
	return fmt.Sprintf("%s:%s", nf.nqnPrefix, expNqnPrefix)
}

func (nf *NameFmt) SecNqnPrefix() string {
	return fmt.Sprintf("%s:%s", nf.nqnPrefix, secNqnPrefix)
}

func (nf *NameFmt) PdBdevName(pdId string) string {
	return fmt.Sprintf("%s-%s", nf.PdBdevPrefix(), pdId)
}

func (nf *NameFmt) PdId(pdBdevName string) string {
	pdId := strings.Replace(pdBdevName, nf.PdBdevPrefix()+"-", "", 1)
	pdId = strings.Replace(pdId, "n1", "", 1) // in the case of Nvme type bdevName
	return pdId
}

func (nf *NameFmt) PdLvsName(pdId string) string {
	return fmt.Sprintf("%s-%s", nf.PdLvsPrefix(), pdId)
}

func (nf *NameFmt) BeLvolName(vdId string) string {
	return fmt.Sprintf("%s", vdId)
}

func (nf *NameFmt) BeLvolFullName(pdId, vdId string) string {
	lvsName := nf.PdLvsName(pdId)
	lvolName := nf.BeLvolName(vdId)
	return fmt.Sprintf("%s/%s", lvsName, lvolName)
}

func (nf *NameFmt) BeNqnName(vdId string) string {
	return fmt.Sprintf("%s-%s", nf.BeNqnPrefix(), vdId)
}

func (nf *NameFmt) FeNqnName(cntlrId string) string {
	return fmt.Sprintf("%s-%s", nf.FeNqnPrefix(), cntlrId)
}

func (nf *NameFmt) FeNvmeName(vdId string) string {
	return fmt.Sprintf("%s-%s", nf.FeNvmePrefix(), vdId)
}

func (nf *NameFmt) FeBdevName(vdId string) string {
	return fmt.Sprintf("%sn1", nf.FeNvmeName(vdId))
}

func (nf *NameFmt) AggBdevName(daId string) string {
	return fmt.Sprintf("%s-%s", nf.AggBdevPrefix(), daId)
}

func (nf *NameFmt) DaLvsName(daId string) string {
	return fmt.Sprintf("%s-%s", nf.DaLvsPrefix(), daId)
}

func (nf *NameFmt) Raid0BdevName(grpId string) string {
	return fmt.Sprintf("%s-%s", nf.Raid0BdevPrefix(), grpId)
}

func (nf *NameFmt) Raid1BdevName(leg0, leg1 string) string {
	return fmt.Sprintf("%s-%s%s", nf.Raid1BdevPrefix(),
		leg0[0:len(leg0)/2], leg1[len(leg1)/2:len(leg1)])
}

func (nf *NameFmt) GrpBdevName(grpId string) string {
	return fmt.Sprintf("%s-%s", nf.GrpBdevPrefix(), grpId)
}

func (nf *NameFmt) SnapshotPrefix() string {
	return "snapshot"
}

func (nf *NameFmt) SnapshotName(snapId string) string {
	return fmt.Sprintf("%s-%s", nf.SnapshotPrefix(), snapId)
}

func (nf *NameFmt) SnapshotFullName(daId, snapId string) string {
	return fmt.Sprintf("%s/%s", nf.DaLvsName(daId), nf.SnapshotName(snapId))
}

func (nf *NameFmt) ClonePrefix() string {
	return "clone"
}

func (nf *NameFmt) CloneName(snapId string) string {
	return fmt.Sprintf("%s-%s", nf.ClonePrefix(), snapId)
}

func (nf *NameFmt) CloneFullName(daId, snapId string) string {
	return fmt.Sprintf("%s/%s", nf.DaLvsName(daId), nf.CloneName(snapId))
}

func (nf *NameFmt) IsClone(bdevName string) bool {
	return strings.Contains(bdevName, nf.ClonePrefix())
}

func (nf *NameFmt) CloneToSnapshot(bdevName string) string {
	return strings.Replace(bdevName, nf.ClonePrefix(),  nf.SnapshotPrefix(), 1)
}

func (nf *NameFmt) ExpNqnName(daName, expName string) string {
	return fmt.Sprintf("%s-%s-%s", nf.ExpNqnPrefix(), daName, expName)
}

func (nf *NameFmt) SecNqnName(cntlrId string) string {
	return fmt.Sprintf("%s-%s", nf.SecNqnPrefix(), cntlrId)
}

func (nf *NameFmt) SecNvmeName(expId string) string {
	return fmt.Sprintf("%s-%s", nf.SecNvmePrefix(), expId)
}

func (nf *NameFmt) ExpId(secNvmeName string) string {
	return strings.Replace(secNvmeName, nf.SecNvmePrefix()+"-", "", 1)
}

func (nf *NameFmt) SecBdevName(expId string) string {
	return fmt.Sprintf("%sn1", nf.SecNvmeName(expId))
}

func (nf *NameFmt) SecNvmeNameToBdevName(secNvmeName string) string {
	return fmt.Sprintf("%sn1", secNvmeName)
}

func (nf *NameFmt) VdSusresPrefix() string {
	return fmt.Sprintf("%s-%s", nf.vdaPrefix, vdSusresType)
}

func (nf *NameFmt) VdSusresName(daId string, grpIdx uint32, vdIdx uint32) string {
	return fmt.Sprintf("%s-%s-%d-%d", nf.VdSusresPrefix(), daId, grpIdx, vdIdx)
}

func (nf *NameFmt) MtNullPrefix() string {
	return fmt.Sprintf("%s-%s", nf.vdaPrefix, mtNullType)
}

func (nf *NameFmt) MtNullName(vdId string) string {
	return fmt.Sprintf("%s-%s", nf.MtNullPrefix(), vdId)
}

func (nf *NameFmt) MtConcatPrefix() string {
	return fmt.Sprintf("%s-%s", nf.vdaPrefix, mtConcatType)
}

func (nf *NameFmt) MtConcatName(vdId string) string {
	return fmt.Sprintf("%s-%s", nf.MtConcatPrefix(), vdId)
}

func (nf *NameFmt) MtRaid1Prefix() string {
	return fmt.Sprintf("%s-%s", nf.vdaPrefix, mtRaid1Type)
}

func (nf *NameFmt) MtRaid1Name(mtId string) string {
	return fmt.Sprintf("%s-%s", nf.MtRaid1Prefix(), mtId)
}

func NewNameFmt(vdaPrefix, nqnPrefix string) *NameFmt {
	return &NameFmt{
		vdaPrefix: vdaPrefix,
		nqnPrefix: nqnPrefix,
	}
}
